api/projects/projects-model.js

```diff
@@ -0,0 +1,66 @@
+const knex = require('knex')
+
+const config = require('../../knexfile')
+
+const db = knex(config.development)
+
+module.exports = {
+    add,
+    attachResource,
+    get,
+    getTree
+}
+
+function add(item) {
+    const {name, description, completed} = item
+    return db('projects').insert({name, description, completed: !!completed}).then(([id]) => get(id))
+}
+
+function get(id) {
+    if (id) {
+        return db('projects')
+        .where({id}).first()
+        .then(resp => {
+            resp.completed = !!resp.completed
+            return resp
+        })
+    }
+    else {
+        return db('projects')
+        .then(resp => resp.map(project => {
+            project.completed = !!project.completed
+            return project
+        }))
+    }
+}
+
+async function getTree(id) {
+    try {
+        const project = await get(id)
+        
+        const tasks = await db('tasks')
+            .select('id', 'description', 'completed')
+            .where('project_id', id)
+            .then(resp => resp.map(task => {
+                task.completed = !!task.completed
+                return task
+            }))
+
+        const resources = await db('project_resources')
+            .join('resources', 'resources.id', 'project_resources.resource_id')
+            .select('resources.*')
+            .where('project_resources.project_id', id)
+
+        return {...project, tasks, resources}
+    }
+    catch (err) {
+        throw err
+    }
+}

```

Comment:
We talked about separation of concerns a bit in our 1:1, and this database method is a good example.

Ideally:
- each database method would only have one database query. 
- The router file would handle the `req` and `res`
- A third file ( in java we call it the 'services' file) would handle calling multiple database methods.

For example, the router for this would look like
```
const projectsService = require("./projects-service")

router.get('/:id', (req, res) => {
    projectsService.getTree(req.params.id)
    .then(resp => {
        res.json(resp)
    })
    .catch(err => {
        console.error(err)
        res.sendStatus(500)
    })
})
```

The Service file would look something like:
```
// Inside the projects-service
const projectsDB = require("./projects-model");

async function getTree(id) {
    try {
        const project = projectsDB.get(id)
        const tasks = projectsDB.getTasksByProject(id)
        const resources = projectsDB.getresourcesByProject(id)
        return {...project, tasks, resources}
    }
    catch (err) {
        throw err
    }
}

module.exports = { getTree }
```

And the projects-model file would have these three methods as separate functions:
```
// Inside the projects-model
function get(id) {
    return db("projects").where({ id }).first();
}

function getTasksByProject(id) {
    return db('tasks')
        .select('id', 'description', 'completed')
        .where('project_id', id)
        .then(resp => resp.map(task => {
            if (task) task.completed = !!task.completed
            return task
        }))
}

function getResourcesByProject(id) {
    return db('project_resources')
        .join('resources', 'resources.id', 'project_resources.resource_id')
        .select('resources.*')
        .where('project_resources.project_id', id)
}

module.exports = { get, getTasksByProject, getResourcesByProject }
```

For a smaller project, this type of separation isn't quite necessary. However, this separation makes testing each individual function much easier - we can test each db query, the router, and the service all on their own without having to integrate anything else.